### ISSUE TITLE
Analyzer Option doesn't work (e. g. for RCS1014)

### ISSUE THREAD

**Issue by anreton** [JosefPihrt/Roslynator#808]
**Product and Version Used**: 
- Microsoft Visual Studio Community 2019 Version 16.10.0
- Roslynator.Analyzers 3.1.0 via Nuget

**Steps to Reproduce**:
1. Add code:
`var numbers = new[] { 42 };`

2. Enable analyzers in .editorconfig:
```
# RCS1014: Avoid implicitly-typed array.
dotnet_diagnostic.RCS1014.severity = warning
dotnet_diagnostic.RCS1014a.severity = warning
dotnet_diagnostic.RCS1014i.severity = warning
```

3. Get false diagnostic for RCS1014 - "Use explicitly typed array".

**Actual Behavior**:
Get false diagnostic for RCS1014 - "Use explicitly typed array".

**Expected Behavior**:
No diagnostic for RCS1014.

**Comment by anreton** [comment]
It looks like it is a duplicate of this [issue](https://github.com/JosefPihrt/Roslynator/issues/757).

**Comment by anreton** [comment]
Install 3.2.0-rc and enable options in .editor config:
```
# RCS1014: Avoid implicitly-typed array.
dotnet_diagnostic.RCS1014.severity = warning
roslynator.RCS1014.use_implicit_type_when_obvious = true
roslynator.RCS1014.invert = true
```
And get again false diagnostic for RCS1014 - "Use explicitly typed array"... (((

**Comment by anreton** [comment]
Aha...
dotnet_diagnostic.* -> roslynator.* in .editorconfig and it all worked. 

```
# RCS1014: Avoid implicitly-typed array.
roslynator.RCS1014.severity = warning
roslynator.RCS1014.use_implicit_type_when_obvious = true
roslynator.RCS1014.invert = true
```

**Comment by JosefPihrt** [comment]
This is correct syntax to specify severity (general syntax not specific to Roslynator):
```
dotnet_diagnostic.RCS1014.severity = warning
```
This is invalid syntax that has no effect on severity:
```
roslynator.RCS1014.severity = warning
```
This is correct syntax how to set options for RCS1014:
```
roslynator.RCS1014.use_implicit_type_when_obvious = true
roslynator.RCS1014.invert = true
```

I think that the problem is that numeric literal is not considered as "obvious" type.

https://github.com/JosefPihrt/Roslynator/blob/67b75f8370c5aa7ca2cf50d208a8eb5adbdc0baf/src/CSharp/CSharp/CSharpTypeAnalysis.cs#L246

If you use for example string literal the code works as expected:
```
var numbers = new[] { "" };
```

**Comment by anreton** [comment]
@JosefPihrt, do I understand correctly that setting options for the analyzer through `dotnet_diagnostic.RCS1014a.severity` does not work in the .editorconfig (according to this [issue](https://github.com/JosefPihrt/Roslynator/issues/804)), but only through by using a ruleset file?

Yes, I achieved the desired behavior only with these settings.
```
dotnet_diagnostic.RCS1014.severity = warning
roslynator.RCS1014.use_implicit_type_when_obvious = false
roslynator.RCS1014.invert = true
```

The problem was that `use_implicit_type_when_obvious` and `invert` were enabled at the same time. That is, if `use_implicit_type_when_obvious` is on, then `invert` does not work, which, of course, has logic.

**Comment by anreton** [comment]
@JosefPihrt, here's another problem with enabling RCS1014.

`new[] { var-of-A, var-of-B }`,

where `A` and `B` have a common base class or implement the same interface.

E. g. I have the following settings for RCS1014:
```
dotnet_diagnostic.RCS1014.severity = warning
roslynator.RCS1014.use_implicit_type_when_obvious = false
roslynator.RCS1014.invert = true
```

And I have false diagnostic `RCS1014: Use implicity typed array` for line:
`new common-type-for-A-and-B[] { var-of-A, var-of-B }`

But I cannot remove the type from the array, because there will be an compiler error [CS0826](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/cs0826).

**Comment by JosefPihrt** [comment]
> @JosefPihrt, do I understand correctly that setting options for the analyzer through e. g. dotnet_diagnostic.RCS1014a.severity does not work in the .editorconfig (according to this issue), but only through by using a ruleset file?

That's correct.

> That is, if use_implicit_type_when_obvious is on, then invert does not work.

That's not true but it makes sense to use only one these options. `invert` simplify invert the behavior of the analyzer whereas `use_implicit_type_when_obvious` changes the behavior in a way that implicit or explicit type is required depending on the "obvious-ness" of the type.

**Comment by anreton** [comment]
@

> > @JosefPihrt, do I understand correctly that setting options for the analyzer through e. g. dotnet_diagnostic.RCS1014a.severity does not work in the .editorconfig (according to this issue), but only through by using a ruleset file?
> 
> That's correct.
> 
> > That is, if use_implicit_type_when_obvious is on, then invert does not work.
> 
> That's not true but it makes sense to use only one these options. `invert` simplify invert the behavior of the analyzer whereas `use_implicit_type_when_obvious` changes the behavior in a way that implicit or explicit type is required depending on the "obvious-ness" of the type.

@JosefPihrt, make sense, yes)

**Comment by anreton** [comment]
> @JosefPihrt, here's another problem with enabling RCS1014.
> 
> `new[] { instance-of-A, instance-of-B }`,
> 
> where `A` and `B` have a common base class or implement the same interface `С`.
> 
> E. g. I have the following settings for RCS1014:
> 
> ```
> dotnet_diagnostic.RCS1014.severity = warning
> roslynator.RCS1014.use_implicit_type_when_obvious = false
> roslynator.RCS1014.invert = true
> ```
> 
> And I have diagnostic `RCS1014: Use implicity typed array` for line:
> `new С[] { instance-of-A, instance-of-B }`
> 
> But I cannot remove the type from the array, because there will be an compiler error [CS0826 - No best type found for implicitly typed array](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/cs0826).

@JosefPihrt, perhaps the option `use_implicit_type_when_not_obvious` for RCS1014 would be helpful for this case.

**Comment by JosefPihrt** [comment]
I'm not getting error from the example you mentioned, If I have following code:
```
interface A
{
    void M()
    {
        var x = new A[] { new B(), new C() };
    }
}

class B : A
{
}

class C : A
{
}
```
after the fix I have:
```
var x = new[] { new B(), (A)new C() };
```
and no error.

> @JosefPihrt, perhaps the new option use_explicit_type_when_not_obvious for RCS1014 would be helpful for this case.

But this option is incorporated in `use_implicit_type_when_obvious`: if the option is enabled and the type is not obvious then the explicit type is required.

**Comment by anreton** [comment]
> I'm not getting error from the example you mentioned, If I have following code:
> 
> ```
> interface A
> {
>     void M()
>     {
>         var x = new A[] { new B(), new C() };
>     }
> }
> 
> class B : A
> {
> }
> 
> class C : A
> {
> }
> ```
> 
> after the fix I have:
> 
> ```
> var x = new[] { new B(), (A)new C() };
> ```
> 
> and no error.

@JosefPihrt, yes, but there is a little extra cast, when there is an opportunity to do without it.

But ok. Thank you so much for your help. Your analyzers are very good, I have been using them for several years. I wish you success.

I think the issue can be closed.

**Comment by JosefPihrt** [comment]
I'm glad that you like Roslynator!
